writer.go

```diff
@@ -18,142 +32,142 @@ type Writer struct {
+				"error establishing index connection to namespace %v: %w",
+				config.Namespace, err)
+		}
+	} else {
+		w.index, err = w.client.Index(host)
+		if err != nil {
+			return nil, fmt.Errorf("error establishing index connection: %w", err)
+		}
 	}
+	sdk.Logger(ctx).Info().Msg("created pinecone index")
 
-	sdk.Logger(ctx).Error().Msgf("metadata: %v", metadata)
-
-	vector := []*pinecone.Vector{{
-		Id:           ID,
-		Values:       payload.Values,
-		SparseValues: payload.PineconeSparseValues(),
-		Metadata:     metadata,
-	}}
+	return &w, nil
+}
 
-	_, err = w.index.UpsertVectors(&ctx, vector)
+func (w *Writer) UpsertVectors(ctx context.Context, vectors []*pinecone.Vector) (int, error) {
+	upserted, err := w.index.UpsertVectors(&ctx, vectors)
 	if err != nil {
-		return fmt.Errorf("error upserting record: %v ", err)
+		return int(upserted), fmt.Errorf("error upserting record: %w", err)
 	}
 
-	sdk.Logger(ctx).Trace().Msg("upserted record")
-	return nil
+	return int(upserted), nil
 }
 
-// Delete deletes records by a key.
-func (w *Writer) Delete(ctx context.Context, record sdk.Record) error {
-	ID := []string{recordID(record.Key)}
-
-	err := w.index.DeleteVectorsById(&ctx, ID)
+func (w *Writer) DeleteVectorsByID(ctx context.Context, vectorIDs []string) error {
+	err := w.index.DeleteVectorsById(&ctx, vectorIDs)
 	if err != nil {
-		return fmt.Errorf("\n error deleting record: %v ", err)
+		return fmt.Errorf("error deleting record: %w", err)
 	}
 
-	sdk.Logger(ctx).Trace().Msg("Successful record delete.")
 	return nil
 }
 
 func (w *Writer) Close() error {
-	return w.index.Close()
-}
-
-// NewPineconeClient takes the Pinecone Index URL string in Config and splits into respective parts to establish a
-// connection
-func NewPineconeClient(ctx context.Context, config DestinationConfig) (*pinecone.Client, *pinecone.IndexConnection, error) {
-	sdk.Logger(ctx).Trace().Msg("Creating a Pinecone Client.")
-
-	client, err := pinecone.NewClient(pinecone.NewClientParams{
-		ApiKey: config.PineconeAPIKey,
-	})
-	if err != nil {
-		return nil, nil, fmt.Errorf("error creating Pinecone client: %v", err)
+	if err := w.index.Close(); err != nil {
+		return fmt.Errorf("error closing writer: %w", err)
 	}
-	sdk.Logger(ctx).Info().Msg("created pinecone client")
-
-	// index urls should have their protocol trimmed
-	hostURL := strings.TrimPrefix(config.PineconeHostURL, "https://")
 
-	index, err := client.Index(hostURL)
-	if err != nil {
-		return nil, nil, fmt.Errorf("error establishing index connection: %v", err)
-	}
-	sdk.Logger(ctx).Info().Msgf("created pinecone index")
+	return nil
+}
 
-	return client, index, nil
+func recordID(key sdk.Data) string {
+	return string(key.Bytes())
 }
 
-func recordID(Key sdk.Data) string {
-	key := Key.Bytes()
-	return string(key)
+type sparseValues struct {
+	Indices []uint32  `json:"indices"`
+	Values  []float32 `json:"values"`
 }
 
 type recordPayload struct {
+	ID           string       `json:"id"`
```

Comment:
This means that a record's payload needs to have the record key as well? We should be able to take it from the record.

Reply:
good catch! that's actually an unneeded field. Commited a delete to that field.

Reply:
I see that the conversation was resolved, but the field is still here. Maybe you didn't push the commit?

Reply:
Weird, I think I did push the commit. Let me look at it

Reply:
Oh yeah did end up not pushing them :sweat_smile: :sweat_smile: